### PR TITLE
fix(server): reject Git-confirmed bare-root projects

### PR DESCRIPTION
**Pull request by walid-baharwal**
## Human hold

This is a narrower project-add policy proposal for [#8164](https://github.com/pingdotgg/t3code/issues/8164), not a complete fix for the original report. Keep it unmerged until a maintainer approves rejecting Git-confirmed bare roots. Already-converted nonbare roots remain accepted. No initialization behavior changes are included.

## Change

A root whose `.git` file points to an embedded bare repository has no checkout of its own. Project add now rejects that root with an explanation and directs the user to an individual worktree. The check applies to CLI add and normalized `project.create` commands.

The original index heuristic rejected a valid unborn separate-git-dir checkout with a linked worktree and accepted a bare repository after `git read-tree --empty` created an index. The correction asks Git for its current bare status instead. It makes one bounded query only for in-root gitdir pointers; ordinary repositories, directories without a pointer, and external linked-worktree pointers keep the no-process fast path. Query errors remain accepted rather than guessing.

The shared pointer parser preserves native separators for the eligibility filter; `devHome` keeps its original backslash normalization. The final query gives Git the `.git` file itself, so Git—not the filter—resolves pointer syntax and trailing spaces. `GIT_WORK_TREE` is removed for this probe, and the explicit gitfile prevents an inherited `GIT_DIR` from redirecting the query.

## Before and after

Real Git, production handlers and disposable SQLite fixtures on Linux. Original guard tested against main [931d41f9](https://github.com/pingdotgg/t3code/commit/931d41f933b2536f699975412169f5318e6adad5). Candidate includes main [2fb99a7a](https://github.com/pingdotgg/t3code/commit/2fb99a7a664faf045f1884f38c620016b34874cb).

| Layout | Original guard | Candidate |
| --- | --- | --- |
| Bare pointer root with linked worktree | Rejected | Rejected |
| Bare pointer root with an index | Incorrectly accepted | Rejected |
| Bare pointer root before its first worktree | Accepted | Rejected |
| Valid unborn separate-git-dir checkout with linked worktree | Incorrectly rejected | Accepted |
| Root already converted to nonbare by `git init` | Rejected by heuristic | Accepted deliberately |
| Valid POSIX `store\name` gitdir beside a different bare `store/name` | Initial query candidate read the wrong path | Correct gitdir queried |

The expanded WorkspacePaths tests fail in four cases against the old guard and pass with the correction. The POSIX backslash regression failed before the raw-parser repair. Three further review regressions failed on [2b1d4d88](https://github.com/pingdotgg/t3code/commit/2b1d4d8837c5428e6559ac2add63dff5bd48c7b2) and pass with the gitfile-query correction: malformed pointer headers beside an unrelated bare repo, a valid trailing-space gitdir beside a bare trimmed-name sibling, and inherited Git environment overrides. Controls cover both `GIT_WORK_TREE` and `GIT_DIR`. CLI/SQLite tests verify that rejected roots create no project record and that valid or already-converted roots remain addable. The same fixtures exercise production command normalization.

```sh
cd apps/server
vp test run src/workspace/WorkspacePaths.test.ts src/bin.test.ts \
  src/orchestration/Normalizer.test.ts src/cli/project.test.ts --maxWorkers 2 \
  -t 'WorkspacePathsLive|current Git bare status|canonicalizeClientCommandTimestamps|maps declared server failures|preserves unexpected server failures'
../../node_modules/.bin/tsgo --noEmit
cd ../../packages/shared
vp test run src/devHome.test.ts src/git.test.ts --maxWorkers 2
../../node_modules/.bin/tsgo --noEmit
```

56 focused tests pass, including the Windows-style `devHome` pointer control. Scoped types, lint, formatting and whitespace checks pass. This is nonvisual server validation; no native macOS client add/start evidence is claimed. At [e80cb32c](https://github.com/pingdotgg/t3code/commit/e80cb32cc6b3b90ebaab473d08eb9309ece707ef), all executed CI jobs, Macroscope Correctness/Approvability and Cursor Bugbot pass. All five review threads are resolved; there are no active change requests. Preview jobs and CodeRabbit's disabled automatic review are skipped. The watcher completed. The human policy hold and missing native add/start proof remain; this is not a complete fix for the original issue.

## Limits

- Git can describe the current repository, not recover the intent of every root already converted to nonbare. A valid unborn checkout and a converted root produced matching bare/inside flags, local config and HEAD in the audit. No replacement heuristic is introduced.
- The exact report's bare-clone commands left `core.bare=true` in the Linux fixture. T3's existing `initRepository` changed it to false, after which the detector accepted the parent and status listed sibling worktrees as untracked. This PR does not change that initialization path or repair existing projects.
- Startup auto-bootstrap dispatches directly and still bypasses command normalization. HTTP dispatch still maps normalization errors to its existing generic `invalid_command` response.
- This does not decide policy for external gitdir pointers, roots that are themselves bare gitdirs, or already-converted layouts. The original issue stays open.

Original implementation by Walid Baharwal is preserved in [48fc1cd9](https://github.com/pingdotgg/t3code/commit/48fc1cd9c46db09ac6ac5ccd79deac3a7e710f50); current main was merged without rewriting the author's history.

Correction and verification: GPT 6 Astra via Codex in T3 Code.

---

> [!NOTE]
> **Medium Risk**
> Changes project-creation validation and shells out to Git for some workspace roots; if Git is unavailable, bare layouts may still be accepted by design.
> 
> **Overview**
> **Project add** now rejects workspace roots whose in-root `.git` file points at an embedded **bare** repository, with a clear error directing users to add an individual worktree instead. The check runs on **`project add`** and on normalized **`project.create`** commands only; other callers still accept bare pointer roots after normalization.
> 
> Detection replaces index/heuristic guards with a bounded **`git rev-parse --is-bare-repository`** query (via **`ensureNotBareRepositoryLayout`**) for eligible in-root gitdir pointers, while keeping a no-Git fast path for ordinary repos and external pointers. Git failures or spawn errors **accept** the root rather than block add.
> 
> Shared **`parseGitDirPointer`** centralizes `.git` file parsing; **`devHome`** uses it for linked-worktree detection. Broad unit and CLI integration tests cover bare, separate-git-dir, worktree, converted, and edge-case layouts.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e80cb32cc6b3b90ebaab473d08eb9309ece707ef. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Reject bare-root Git repositories in `project-add` and command normalization
> - Adds `ensureNotBareRepositoryLayout` to the `WorkspacePaths` service, which probes when `.git` is a gitfile pointing strictly inside the normalized root and asks Git `--is-bare-repository`; nonzero or failed Git results are treated as not bare.
> - CLI `projectAddMutation` in [project.ts](https://github.com/pingdotgg/t3code/pull/8646/files#diff-44f1264ca9e5b2f3994c44291b5616390842fc6345021ca54e8447fd8dbbc1cd) and orchestration `normalizeDispatchCommand` in [Normalizer.ts](https://github.com/pingdotgg/t3code/pull/8646/files#diff-751e99c6b2a5986d684c4717507023bfb24d6250386e071d6b48dd195e10c5ea) now call the check after `normalizeWorkspaceRoot`, returning a new `WorkspaceRootBareRepositoryLayoutError` (mapped to `OrchestrationDispatchCommandError` in the normalizer).
> - Extracts a shared `parseGitDirPointer` utility in [git.ts](https://github.com/pingdotgg/t3code/pull/8646/files#diff-e52ee60f55e13d4ce4a00279168044882e11385edf15a01ad8073cd1442f5785) for `gitdir:` lines from `.git` files; `devHome.pointsAtLinkedWorktree` now uses it.
> - Behavioral Change: roots classified as in-root bare repositories are now rejected at add/normalize time instead of being accepted as projects; Git spawn failures or non-`true` results do not trigger rejection.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized e80cb32.</sup>
> 
>

### CODE REVIEW

**Comment by juliusmarminge** [comment]
Current-main audit for [#8164](https://github.com/pingdotgg/t3code/issues/8164), tested on [0dd5c64bc](https://github.com/pingdotgg/t3code/commit/0dd5c64bcfac7974d81bac76740ecd83540077e8) on September 4, 2026.

The exact bare-clone, `.git` pointer, and sibling-worktree commands leave `core.bare=true` in my disposable Linux fixture. The production detector returns null for the parent and detects the child worktree correctly. Calling the production `initRepository` on that parent changes `core.bare` to false. The detector then accepts the parent, and Git status reports the sibling worktrees as untracked. That reproduces the problematic layout, with an initialization step beyond the report's shell commands.

This PR still needs human review. The missing-index heuristic has outstanding counterexamples in the existing reviews, including a legitimate empty separate-git-dir repository. I have not verified the current PR head against those cases or the actual add/start UI path, so this comment is not a merge-readiness recommendation. Please keep the valid-layout controls when revising the classification rule.

GPT 6 Astra via Codex in T3 Code.

File: apps/server/src/cli/project.ts [hunk 1]
```diff
@@ -461,7 +461,10 @@ const projectAddCommand = Command.make("add", {
           command: ProjectCliDispatchCommand,
         ) => Effect.Effect<void, Error, FileSystem.FileSystem | HttpClient.HttpClient | Path.Path>;
       }) {
-        const workspaceRoot = yield* normalizeWorkspaceRootForProjectCommand(flags.workspaceRoot);
+        const workspacePaths = yield* WorkspacePaths.WorkspacePaths;
+        const workspaceRoot = yield* normalizeWorkspaceRootForProjectCommand(
+          flags.workspaceRoot,
+        ).pipe(Effect.flatMap(workspacePaths.ensureNotBareRepositoryLayout));
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/server/src/cli/project.ts:466]: Rechecked the current [48fc1cd9](https://github.com/pingdotgg/t3code/commit/48fc1cd9c46db09ac6ac5ccd79deac3a7e710f50) patch rebased locally onto main [82f64cd8](https://github.com/pingdotgg/t3code/commit/82f64cd8d046ab4ee8588f7bcbc1e66a4a0c80fe), using two real-Git fixtures through production WorkspacePaths. Both fail the expected classification. A plain fresh separate-git-dir checkout is now accepted because of the worktrees guard, so that part of the older finding needs narrowing. After `git worktree add --orphan -b feature <sibling>` it is still a valid unborn working tree with core.bare=false and no index, but the new guard rejects it. Conversely, a bare container with an orphan linked worktree is initially rejected, then `git --git-dir <root>/.bare read-tree --empty` creates a valid index without changing its bare status, and the guard accepts it. Git independently reports --is-inside-work-tree=true for the first fixture and --is-bare-repository=true for the second. Exact command: `vp test run apps/server/src/workspace/BareLayoutAudit.test.ts --maxWorkers=2` in the isolated current-main integration. No live repositories or app data were changed. The index heuristic remains a fix-first blocker; detecting a still-bare repo from Git metadata and deciding what to do with already-converted ambiguous layouts are separate choices. GPT 6 Astra via Codex in T3 Code.

[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/server/src/cli/project.ts:466]: The index heuristic is replaced in [b89cfd61](https://github.com/pingdotgg/t3code/commit/b89cfd618f30f55006bc0dac4f63db4b17f38430), included in current head [2b1d4d88](https://github.com/pingdotgg/t3code/commit/2b1d4d8837c5428e6559ac2add63dff5bd48c7b2).

Real Git regressions now accept the valid unborn separate-git-dir checkout with an orphan linked worktree and reject the still-bare root after `read-tree --empty` creates its index. The expanded WorkspacePaths tests failed four cases with the previous guard and pass with the correction. A further real-Git control caught a literal POSIX backslash being normalized into a different gitdir; the parser now returns the raw target and `devHome` retains its previous normalization.

The final main integration passes 53 focused tests, server/shared types, targeted lint and formatting. Four actual CLI/SQLite fixtures also exercise command normalization and verify rejected roots create no project record. Current-head CI is running.

This resolves the concrete index-classification findings, not the whole policy decision. The PR is explicitly HUMAN HELD: only Git-confirmed still-bare in-root pointers are rejected; already-converted nonbare roots remain accepted. Initialization, startup auto-bootstrap and native macOS add/start behavior are not changed or claimed verified. [#8164](https://github.com/pingdotgg/t3code/issues/8164) remains open.

GPT 6 Astra via Codex in T3 Code.

File: packages/shared/src/git.ts [hunk 2]
```diff
@@ -10,6 +10,20 @@ import * as Arr from "effect/Array";
...
+ * pointer. Callers apply the path semantics appropriate to their use.
+ */
+export const parseGitDirPointer = (gitFileContents: string): string | undefined => {
+  const gitdir = gitFileContents
+    .split(/\r?\n/)
+    .map((line) => line.trim())
+    .find((line) => line.startsWith("gitdir:"))
+    ?.slice("gitdir:".length)
+    .trim();
+  return gitdir === undefined || gitdir.length === 0 ? undefined : gitdir;
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply packages/shared/src/git.ts:24]: Fixed in [e80cb32c](https://github.com/pingdotgg/t3code/commit/e80cb32cc6b3b90ebaab473d08eb9309ece707ef) by passing the `.git` file itself to Git, rather than passing the filter's parsed target. The existing helper remains an eligibility filter, preserving `devHome` behavior; Git now owns gitfile syntax and path interpretation.

Actual Git regressions reproduce both failures on the previous head: malformed metadata/header variants could query an unrelated bare directory, and a valid `store ` checkout could query a different bare `store`. With the correction, malformed gitfiles exit nonzero and are accepted by the existing fail-open rule; the valid trailing-space checkout is accepted using its actual gitdir. The literal POSIX backslash control still passes.

The three new review regression tests failed on [2b1d4d88](https://github.com/pingdotgg/t3code/commit/2b1d4d8837c5428e6559ac2add63dff5bd48c7b2) and pass with this commit. All 56 scoped tests, server/shared typechecks, targeted lint and formatting pass. Current-head CI is running. This resolves the concrete misclassification, not the PR's explicit human policy hold.

GPT 6 Astra via Codex in T3 Code.

File: apps/server/src/workspace/WorkspacePaths.ts [hunk 3]
```diff
@@ -149,6 +173,47 @@ export const make = Effect.gen(function* () {
...
+      gitDirFromRoot.length === 0 ||
+      gitDirFromRoot === ".." ||
+      gitDirFromRoot.startsWith(`..${path.sep}`) ||
+      path.isAbsolute(gitDirFromRoot)
+    ) {
+      return false;
+    }
+
+    return yield* vcsProcess
+      .run({
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/server/src/workspace/WorkspacePaths.ts:202]: Fixed in [e80cb32c](https://github.com/pingdotgg/t3code/commit/e80cb32cc6b3b90ebaab473d08eb9309ece707ef): this probe supplies `GIT_WORK_TREE: undefined`, which the existing process runner removes from the child environment.

The regression uses actual Git subprocesses: the bare fixture reports false with the inherited worktree override on the previous head, but the corrected service rejects it. A second control sets `GIT_DIR` to another nonbare repository; the explicit `--git-dir <root>/.git` query still examines the intended root. No global process environment or real repositories were changed.

The three new review regression tests failed before this correction; all 56 focused tests plus server/shared types, lint and formatting now pass. Current-head CI is running. The narrowed project-add policy remains HUMAN HELD and this does not change initialization or already-converted-root behavior.

GPT 6 Astra via Codex in T3 Code.

**Comment by juliusmarminge** [comment]
Thanks for the PR. We're not taking changes to the orchestration and provider layers right now: that part of the server is being rewritten for V2, and merging into the current code would either conflict with or be thrown away by that work.

Closing for now. If this is still an issue once V2 lands, please reopen (or open a fresh PR against the new code) and we'll take a proper look.
